### ISSUE TITLE
Something completely wrong with reference assignment in bytecode

### ISSUE THREAD

**Issue by artemkin** [janestreet/core_kernel#40]
The program prints "ERROR" instead of "OK".
It prints "OK", as expected, if you remove `open Core_kernel.Std` or build native version of the program.

Reproducible on both Mac OS X 10.10 and CentOS 7.

```
airbook:bug stas$ ocamlc -version
4.02.1
airbook:bug stas$ opam list | grep core
core               112.24.00  Industrial strength alternative to OCaml's standar
core_extended      112.24.00  Extra components that are not as closely vetted or
core_kernel        112.24.00  Industrial strength alternative to OCaml's standar
airbook:bug stas$ cat _tags
true: package(core_kernel)
true: thread,debug,annot
true: bin_annot

airbook:bug stas$ cat bug.ml

open Core_kernel.Std

let _ =
  let cont = ref (fun x -> Printf.printf "ERROR\n"; x) in
  (cont := (fun x -> Printf.printf "OK\n"; x); ())
  |> !cont

airbook:bug stas$ ocamlbuild -use-ocamlfind bug.byte && ./bug.byte
Finished, 3 targets (3 cached) in 00:00:00.
ERROR
airbook:bug stas$
```

**Comment by yminsky** [comment]
I don't believe this is a bug in Core_kernel.

I'm not sure where the difference between core_kernel's |> and the stdlib's, but the code here depends in an unfortunate way on order of evaluation.  Roughly, I think the code itself is buggy.  The following code is more robust:

    open Core_kernel.Std

    let _ =
      let cont = ref (fun x -> Printf.printf "ERROR\n"; x) in
      (cont := (fun x -> Printf.printf "OK\n"; x); ())
      |> (fun () -> !cont)

basically, if you think of |> as a function, then you should not depend on the order of evaluation of its arguments (which is generally undefined in OCaml.)  The variation that I wrote doesn't depend on order of evaluation, since it explicitly deferes the dereferencing of the ref to the right hand side of a lambda.

**Comment by yminsky** [comment]
Also, there's something weird with the stdlib on this point.  Note that this prints OK

    let _ =
      let cont = ref (fun x -> Printf.printf "ERROR\n"; x) in
      (cont := (fun x -> Printf.printf "OK\n"; x); ())
      |> !cont

But this prints ERROR:

    let (|>) = Pervasives.(|>)

    let _ =
      let cont = ref (fun x -> Printf.printf "ERROR\n"; x) in
      (cont := (fun x -> Printf.printf "OK\n"; x); ())
      |> !cont

That's effectively what Core_kernel is doing to get its own version of |>.

**Comment by diml** [comment]
The difference between core_kernel's |> and the stdlib's is how it is defined. In `Pervasives` it is defined as:

```ocaml
external ( |> ) : 'a -> ('a -> 'b) -> 'b = "%revapply"
```

and is optimised away at an early stage of the compilation, in the common path for byte and native code.

In core_kernel it is defined as a classic function:

```ocaml
val (|>) : 'a -> ('a -> 'b) -> 'b
```

In general projects using core also use the pa_pipebang syntax extension which expands `|>` at parsing time.

**Comment by artemkin** [comment]
Wow! I didn't think I could experience that classic C-style problem in OCaml )

Thank you

**Comment by yminsky** [comment]
@diml, I believe it's defined as rev-apply for Core_kernel too.  Somehow, this behavior changes even with just pervasives when you rebind |>, which seems insane.

Here's the Core_kernel implementation:

https://github.com/janestreet/core_kernel/blob/master/lib/fn.mli#L7

Is there a bug to report upstream to the compiler?

**Comment by diml** [comment]
@yminsky, I think the one you get when you open `Core_kernel.Std` comes from `common.mli` where it is not defined as an external.

About the rebinding, this is the same as when you rebind `Pervasives.compare`. Due to the way it is done in the compiler, to get the specialised code you really have to reference the value that is defined as an external.

**Comment by yminsky** [comment]
Sure, but I am surprised about the semantic change.  I normally think of
that issue as performance-only.  Having a semantic change fit in this way
feels like a bug to me.

**Comment by bmillwood** [comment]
The semantics in both cases is "the arguments to |> will be evaluated in some unspecified order", right? :)

**Comment by yminsky** [comment]
Fair point!
